### PR TITLE
Added Markdown Editor

### PR DESCRIPTION
**Pull request by EOEboh**
> First thing, PLEASE READ THIS: [ReactPlay Code Review Checklist](https://github.com/reactplay/react-play/wiki/ReactPlay-Code-Review-Checklist)

# Description
I added a new play called Markdown Editor #430 .

It's a beginner-level play which displays input texts as markdown output. It utilizes a dependency called 'remarkable'.

My main motivation is to teach beginners the React library by making them take baby steps before tackling advanced concepts. 

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

It works perfectly on localhost, netlify, vercel and mobile platforms.

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Sachin-chaurasiya** [comment]
Hey @EOEboh , Thanks for the PR.

In the code, I can see you are using `dangerouslySetInnerHTML` within the div tag.
```
<div className='output-div' dangerouslySetInnerHTML={{__html: md.render(text)}} >
```
Can you avoid this?
Or can we use [tui editor](https://ui.toast.com/tui-editor/) , which will provide both markdown and WYSIWYG mode?

Tui Editor is highly configurable and customisable, with plugin architecture it allows you to add your own plugin and also you can use an existing plugin to add an extra feature to the markdown editor.

image

cc : @atapas

**Comment by EOEboh** [comment]
> Hey @EOEboh , Thanks for the PR.
> 
> In the code, I can see you are using `dangerouslySetInnerHTML` within the div tag.
> 
> ```
> <div className='output-div' dangerouslySetInnerHTML={{__html: md.render(text)}} >
> ```
> 
> Can you avoid this? Or can we use [tui editor](https://ui.toast.com/tui-editor/) , which will provide both markdown and WYSIWYG mode?
> 
> Tui Editor is highly configurable and customisable, with plugin architecture it allows you to add your own plugin and also you can use an existing plugin to add an extra feature to the markdown editor.
> 
> image
> 
> cc : @atapas

Thanks @Sachin-chaurasiya 
I would check it out.

Quick one: I did a `git reset` yesterday and I couldn't find my play on localhost after pushing changes. Do you have any suggestions how to fix it

**Comment by EOEboh** [comment]
@atapas can you review this please?

**Comment by atapas** [comment]
> > Hey @EOEboh , Thanks for the PR.
> > In the code, I can see you are using `dangerouslySetInnerHTML` within the div tag.
> > ```
> > <div className='output-div' dangerouslySetInnerHTML={{__html: md.render(text)}} >
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Can you avoid this? Or can we use [tui editor](https://ui.toast.com/tui-editor/) , which will provide both markdown and WYSIWYG mode?
> > Tui Editor is highly configurable and customisable, with plugin architecture it allows you to add your own plugin and also you can use an existing plugin to add an extra feature to the markdown editor.
> > image
> > cc : @atapas
> 
> Thanks @Sachin-chaurasiya I would check it out.
> 
> Quick one: I did a `git reset` yesterday and I couldn't find my play on localhost after pushing changes. Do you have any suggestions how to fix it

@EOEboh Do you still have the issue?

**Comment by atapas** [comment]
@EOEboh Could you please resolve merge conflicts?

**Review by atapas — CHANGES_REQUESTED** [review]
Please do not commit this file

src/plays/index.js

File: src/plays/markdown-editor/Readme.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Markdown Editor
+
```
Comment (atapas) [inline src/plays/markdown-editor/Readme.md:2]: Please add more details like,

- What someone will learn from this project?
- You have included an NPM, please mention that in the readme.

File: src/plays/markdown-editor/styles.css [hunk 2]
```diff
@@ -0,0 +1,26 @@
+/* enter stlyes here */
+*{
```
Comment (atapas) [inline src/plays/markdown-editor/styles.css:2]: Do not override with *

File: src/plays/markdown-editor/styles.css [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    background-color: #724040;
+  }
+  .main{
+    display: flex;
+    flex-direction: column;
+    text-align: center;
+    
+  }
+  
+  h1, h3{
```
Comment (atapas) [inline src/plays/markdown-editor/styles.css:16]: Do not override selectors.

File: src/plays/markdown-editor/styles.css [hunk 4]
```diff
@@ -0,0 +1,26 @@
+/* enter stlyes here */
+*{
+    box-sizing: border-box; 
+  }
+  
+  .play-details{
```
Comment (atapas) [inline src/plays/markdown-editor/styles.css:6]: This is a common style across plays.. Please add a unique style and use that.

File: src/plays/markdown-editor/styles.css [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    background-color: #724040;
+  }
+  .main{
```
Comment (atapas) [inline src/plays/markdown-editor/styles.css:9]: This is has got higher chances of getting overridden. Please use a unique class name.

**Review by atapas — CHANGES_REQUESTED** [review]
@EOEboh 

Provided my review comments

File: src/plays/index.js [hunk 6]
```diff
@@ -1,5 +1,4 @@
 // Do not modify the content of this file
```
Comment (atapas) [inline src/plays/index.js:1]: @EOEboh as mentioned this file shouldn't be committed.

File: src/plays/markdown-editor/MarkdownEditor.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+function MarkdownEditor(props) {
+
+  // Your Code Start below.
+
+   // state for textarea input value
+   const [ text, setText ] = useState("");
+
+  return (
+    <>
+      <div className="play-details">
```
Comment (atapas) [inline src/plays/markdown-editor/MarkdownEditor.js:19]: Add a unique class name here along with `play-details`. Like 

```html
<div className="eboh-md-editor play-details">
```

Now use this one as a prefix to the styles in your CSS file. Like
```css
.eboh-md-editor.play-details {


} 

.eboh-md-editor .wrapper-div {

}
// Like the above.
```

File: src/plays/markdown-editor/MarkdownEditor.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            <label htmlFor='markdown'><h3 className='heading-3'>Type Here </h3></label> 
+             <textarea cols='30' rows='10' id='markdown' name='markdown' placeholder='Type Something...' required
+            value={text}
+            onChange={(e) => setText(e.target.value)}
+          >
+            </textarea>
+          
+        <h3 className='heading-3'> Preview</h3>
+
+          <div className='output-div' dangerouslySetInnerHTML={{__html: md.render(text)}} >
```
Comment (atapas) [inline src/plays/markdown-editor/MarkdownEditor.js:35]: Do you need the `Remarkable`? As I understand, we are not highlighting any syntax here, so why do we need to treat the output as HTML, why not the plain text?

File: src/plays/markdown-editor/styles.css [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+  
+  
+  .play-details{
+    background-color: #724040;
+    font-family: system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
+    box-sizing: var(--box);
+  }
+  
+  .wrapper-div{
```
Comment (atapas) [inline src/plays/markdown-editor/styles.css:14]: Please follow the structure I mentioned about the unique style name.

File: src/plays/markdown-editor/styles.css [hunk 10]
```diff
@@ -0,0 +1,36 @@
+/* enter stlyes here */
+:root {
```
Comment (atapas) [inline src/plays/markdown-editor/styles.css:2]: Please do not override. These selectors are already overridden at the framework root level. So remove.

**Review by atapas — CHANGES_REQUESTED** [review]
@EOEboh A few more change requests.

**Comment by atapas** [comment]
@EOEboh 

Also please check this error. The build fails due to this



static/css/main.617be5a3.css from Css Minimizer plugin
--
08:04:26.826 | /vercel/path0/static/css/main.617be5a3.css:7170:27: Missed semicolon [webpack://./src/plays/markdown-editor/styles.css:11,4][static/css/main.617be5a3.css:7170,27]

**Comment by EOEboh** [comment]
@atapas I have made corrections.
However, the conflicting `src/plays/index.js` keeps being pushed despite being in the `.gitignore` file.

Please can you resolve the conflict? Or tell me what to do.

**Comment by EOEboh** [comment]
@koustov could you have a look, if it's good to go

**Comment by atapas** [comment]
> @koustov could you have a look, if it's good to go

To me it's almost there. But you need to resolve the conflicts. Just revert the file content back to whatever is in the main.

File: src/plays/markdown-editor/styles.css [hunk 11]
```diff
@@ -0,0 +1,33 @@
+/* enter stlyes here */
+  
+  .md-editor.play-details{
+    background-color: #724040;
+    font-family: system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
+    box-sizing: border-box;
+  }
+
```
Comment (atapas) [inline src/plays/markdown-editor/styles.css:8]: use `.md-editor` for rest of the styles too.

**Review by atapas — CHANGES_REQUESTED** [review]

**Comment by atapas** [comment]
@EOEboh Please take a look at my comments.

**Comment by EOEboh** [comment]
> > @koustov could you have a look, if it's good to go
> 
> To me it's almost there. But you need to resolve the conflicts. Just revert the file content back to whatever is in the main.

What do you really mean here? Can you outline the steps?

**Comment by koustov** [comment]
I have reviewed and it looks good.
@EOEboh a play got merged, so you need to resolve the conflicts in order to merge

**Review by koustov — APPROVED** [review]
lgtm

**Comment by EOEboh** [comment]
> I have reviewed and it looks good. @EOEboh a play got merged, so you need to resolve the conflicts in order to merge

I have resolved it now. Please confirm

**Comment by koustov** [comment]
> > I have reviewed and it looks good. @EOEboh a play got merged, so you need to resolve the conflicts in order to merge
> 
> I have resolved it now. Please confirm

Approved.

**Comment by EOEboh** [comment]
> > > I have reviewed and it looks good. @EOEboh a play got merged, so you need to resolve the conflicts in order to merge
> > 
> > 
> > I have resolved it now. Please confirm
> 
> Approved.

Feels great to see this lol

**Comment by koustov** [comment]
Hey @EOEboh , this looks good however do you feel like updating the UI a bit to make it more contemporary? 

You can refer 
- https://dillinger.io/
- https://markdownlivepreview.com/
- https://codebeautify.org/markdown-editor
- https://www.tutorialspoint.com/online_markdown_editor.php

**Comment by EOEboh** [comment]
> Hey @EOEboh , this looks good however do you feel like updating the UI a bit to make it more contemporary?
> 
> You can refer
> 
> * https://dillinger.io/
> * https://markdownlivepreview.com/
> * https://codebeautify.org/markdown-editor
> * https://www.tutorialspoint.com/online_markdown_editor.php

Sure. I'd work on that

**Comment by EOEboh** [comment]
@koustov and @atapas  I have modified the UI
I also added a download feature, to download the HTML text output as a pdf file

**Comment by atapas** [comment]
Thanks for improving it @EOEboh 

Can you please change the text colour of the type here and preview the texts? They are barely readable with black font and brown background. This is a mandatory fix.

Other optional improvement suggestions:

- The boxes can be side by side, then up and down. They should be responsive so that on a small screen, they align column-wise.
- Can provide border-radius to the boxes.
- The Download button can be disabled if there are no texts.
- The cover image design can be better.

Please note that the above comments are optional for you to take care of, but if you do, it will help the play to perceive well. Also, if you think of getting it done and need any help or peer-programming, just let us know in the discord. One of us will work with you.

Thanks.

**Comment by EOEboh** [comment]
> 

Okay, I'd work on it

**Comment by EOEboh** [comment]
@atapas I have made corrections and implemented some of your suggestions

**Comment by koustov** [comment]
> > 
> 
> @atapas Okay, I'd work on it. I just wrote an article about the play, is there a way I can edit the description, and link it?

You can create a link on your play

**Comment by koustov** [comment]
> @atapas I have made corrections and implemented your suggestions

Hey @EOEboh , me and @atapas  were suggesting on similar lines. Few things are still pending
- A better cover page
- `Type Here` and `Preview` are in black on top of dark back ground which is not readable
- Disable download button when not text in `preview` box
- Make both horizontally aligned and make the most of real estate. something like 
image


Could you please make the changes

**Comment by EOEboh** [comment]
> > @atapas I have made corrections and implemented your suggestions
> 
> Hey @EOEboh , me and @atapas were suggesting on similar lines. Few things are still pending
> 
> * A better cover page
> * `Type Here` and `Preview` are in black on top of dark back ground which is not readable
> * Disable download button when not text in `preview` box
> * Make both horizontally aligned and make the most of real estate. something like
>   image
> 
> Could you please make the changes

I already implemented them 🙂

**Comment by koustov** [comment]
> > > @atapas I have made corrections and implemented your suggestions
> > 
> > 
> > Hey @EOEboh , me and @atapas were suggesting on similar lines. Few things are still pending
> > 
> > * A better cover page
> > * `Type Here` and `Preview` are in black on top of dark back ground which is not readable
> > * Disable download button when not text in `preview` box
> > * Make both horizontally aligned and make the most of real estate. something like
> >   image
> > 
> > Could you please make the changes
> 
> I already implemented them 🙂

😮 then I am surprised because I didn't see it. Will check again

**Comment by koustov** [comment]
@EOEboh , looks like latest build didn't get deployed to vercel, it still shows the thumbnail as 
image

and the play as
image

Will wait for @atapas to fix the vercel issue first

**Comment by atapas** [comment]
@EOEboh much better looking, great work

@koustov, please conclude.

**Comment by koustov** [comment]
UI is impressive @EOEboh. I will finish my review by EOD

**Comment by EOEboh** [comment]
> UI is impressive @EOEboh. I will finish my review by EOD

Thank you @koustov and @atapas .. this means a lot to me.

File: package.json [hunk 12]
```diff
@@ -19,7 +19,9 @@
     "git-repo-api": "^0.0.17",
     "graphql": "^16.5.0",
     "html-to-image": "^1.9.0",
-    "json-graphql-parser": "^0.1.7",
+    "html2canvas": "^1.4.1",
+    "json-graphql-parser": "^0.0.20",
```
Comment (koustov) [inline package.json:23]: Let's not downgrade the library.

File: src/plays/markdown-editor/Downloader.jsx [hunk 13]
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+
+// html2canvas for converting the md to png image
+import html2canvas from 'html2canvas';
```
Comment (koustov) [inline src/plays/markdown-editor/Downloader.jsx:4]: Lets not use `html2canvas` as it brings the style as well

Comment (koustov) [inline src/plays/markdown-editor/Downloader.jsx:4]: also, as its a png/jpg, one cannot copy the text from pdf

Comment (koustov) [reply src/plays/markdown-editor/Downloader.jsx:4]: image

File: src/plays/markdown-editor/MarkdownEditor.js [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+        Preview
+      </h3> */}
+
+          <Output 
+          md={md}
+          text={text}
+          elementId='elementId'
+          />
+        </div>
+          <Downloader
```
Comment (koustov) [inline src/plays/markdown-editor/MarkdownEditor.js:50]: Need to disable it when no text is rendered. I could download an empty document.

File: src/plays/markdown-editor/MarkdownEditor.js [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+             <textarea cols='40' rows='10' id='markdown'    name='markdown' placeholder='Type Markdown...' required
+              value={text}
+              onChange={(e) => setText(e.target.value)}
+            >
+            
+            </textarea>
+          
+            {/* <h3 className='md-editor heading-3'>
+        Preview
+      </h3> */}
```
Comment (koustov) [inline src/plays/markdown-editor/MarkdownEditor.js:42]: Remove commented snippets

File: src/plays/markdown-editor/MarkdownEditor.js [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+          <Output 
+          md={md}
+          text={text}
+          elementId='elementId'
```
Comment (koustov) [inline src/plays/markdown-editor/MarkdownEditor.js:47]: it's very generic id, I would suggest using more specific one like `mde-preview-box`

File: src/plays/markdown-editor/Readme.md [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+- Language: js
+- Level: Beginner
+
+## Creator Information
+
+- User: EOEboh
+- Gihub Link: https://github.com/EOEboh
+- Blog: 
+- Video:
```
Comment (koustov) [inline src/plays/markdown-editor/Readme.md:15]: Lets remove unused properties

File: src/plays/markdown-editor/styles.css [hunk 18]
```diff
@@ -0,0 +1,68 @@
+/* enter stlyes here */
```
Comment (koustov) [inline src/plays/markdown-editor/styles.css:1]: Can you run `prettier` on this file? Looks like the alignments are not proper

**Review by koustov — CHANGES_REQUESTED** [review]
Few observations

**Comment by koustov** [comment]
@EOEboh is it ready for final review?

**Comment by EOEboh** [comment]
> @EOEboh is it ready for final review?

@koustov I'm sorry I'm delaying, I was down these few days. I would work on the changes today. I apologize

**Comment by EOEboh** [comment]
@koustov I fixed the issues.

**Review by koustov — APPROVED** [review]
lgtm

**Comment by EOEboh** [comment]
> lgtm

@koustov I made a small change.
I think everything is set.

**Review by atapas — APPROVED** [review]

**Comment by atapas** [comment]
@all-contributors please add @EOEboh for Code

**Pull request merged**